pkg/handlers/nutanix/mutation/controlplanefailuredomains/inject.go

```diff
@@ -0,0 +1,122 @@
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/patches/selectors"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/variables"
+)
+
+const (
+	// VariableName is the external patch variable name.
+	VariableName = "failureDomains"
+)
+
+type nutanixControlPlaneFailureDomains struct {
+	variableName      string
+	variableFieldPath []string
+}
+
+func NewPatch() *nutanixControlPlaneFailureDomains {
+	return newNutanixControlPlaneFailureDomains(
+		v1alpha1.ClusterConfigVariableName,
+		v1alpha1.ControlPlaneConfigVariableName,
+		v1alpha1.NutanixVariableName,
+		VariableName,
+	)
+}
+
+func newNutanixControlPlaneFailureDomains(
+	variableName string,
+	variableFieldPath ...string,
+) *nutanixControlPlaneFailureDomains {
+	return &nutanixControlPlaneFailureDomains{
+		variableName:      variableName,
+		variableFieldPath: variableFieldPath,
+	}
+}
+
+func (h *nutanixControlPlaneFailureDomains) Mutate(
+	ctx context.Context,
+	obj *unstructured.Unstructured,
+	vars map[string]apiextensionsv1.JSON,
+	holderRef runtimehooksv1.HolderReference,
+	_ client.ObjectKey,
+	_ mutation.ClusterGetter,
+) error {
+	log := ctrl.LoggerFrom(ctx).WithValues(
+		"holderRef", holderRef,
+		"variableName", h.variableName,
+		"variableFieldPath", h.variableFieldPath,
+	)
+
+	controlPlaneFDsVar, err := variables.Get[[]string](
+		vars,
+		h.variableName,
+		h.variableFieldPath...,
+	)
+	if err != nil {
+		if variables.IsNotFoundError(err) {
+			log.Info("ControlPlane nutanix failureDomains variable not defined", "error", err.Error())
+			return nil
+		}
+		log.Error(err, "failed to get controlPlane nutanix failureDomains variable.")
+		return err
+	}
+
+	log = log.WithValues("variableValue", controlPlaneFDsVar)
+
+	if len(controlPlaneFDsVar) == 0 {
+		return nil
+	}
+
+	return patches.MutateIfApplicable(
+		obj,
+		vars,
+		&holderRef,
+		selectors.InfrastructureCluster(capxv1.GroupVersion.Version, "NutanixClusterTemplate"),
+		log,
+		func(obj *capxv1.NutanixClusterTemplate) error {
+			log.WithValues(
+				"patchedObjectKind", obj.GetObjectKind().GroupVersionKind().String(),
+				"patchedObjectName", client.ObjectKeyFromObject(obj),
+			).Info("setting controlPlaneFailureDomains in NutanixCluster spec")
+
+			fdRefs := []corev1.LocalObjectReference{}
+			// use to check duplicates
+			fdSet := map[string]bool{}
+			for _, fd := range controlPlaneFDsVar {
+				if _, ok := fdSet[fd]; ok {
+					log.Error(fmt.Errorf("duplicate failureDomain configured: %s", fd), "")
+					continue
+				}
+				fdSet[fd] = true
+				fdRefs = append(fdRefs, corev1.LocalObjectReference{Name: fd})
+			}
+
+			// set controlPlaneFailureDomains in NutanixCluster spec
+			obj.Spec.Template.Spec.ControlPlaneFailureDomains = fdRefs
+			log.Info("done setting controlPlaneFailureDomains in NutanixClusterTemplate spec",
```

Comment:
I think its ok to drop this log to be consistent with the other handlers.

Reply:
removed